compiler/src/dotty/tools/dotc/transform/SpecializeFunction1.scala

```diff
@@ -0,0 +1,164 @@
+    case _ => ref
+  }
+
+  override def transformTemplate(tree: Template)(implicit ctx: Context, info: TransformerInfo) =
+    tree match {
+      case ShouldTransformTree(func1, t1, r) => transformTree(tree, func1, t1, r)
+      case _ => tree
+    }
+
+  private[this] val functionName = "JFunction1".toTermName
+  private[this] val functionPkg  = "scala.compat.java8.".toTermName
+  private[this] var Function1: Symbol = _
+  private[this] var argTypes: Set[Symbol] = _
+  private[this] var returnTypes: Set[Symbol] = _
+  private[this] var blacklisted: Set[Symbol] = _
+
+  /** Do setup of `argTypes` and `returnTypes` */
+  override def prepareForUnit(tree: Tree)(implicit ctx: Context) = {
+    argTypes = Set(defn.DoubleClass,
+                   defn.FloatClass,
+                   defn.IntClass,
+                   defn.LongClass,
+                   defn.UnitClass,
+                   defn.BooleanClass)
+
+    returnTypes = Set(defn.DoubleClass,
+                      defn.FloatClass,
+                      defn.IntClass,
+                      defn.LongClass)
+
+    Function1 = ctx.requiredClass("scala.Function1")
+
+    blacklisted = Set(
+      "scala.compat.java8.JFunction1",
+      "scala.runtime.AbstractFunction1",
+      "scala.PartialFunction",
+      "scala.runtime.AbstractPartialFunction"
+    ).map(ctx.requiredClass(_))
+
+    this
+  }
+
+  private object ShouldTransformDenot {
+    def unapply(cref: ClassDenotation)(implicit ctx: Context): Option[(ClassDenotation, Type, Type, Type)] = {
+      def collectFunc1(xs: List[Type])(implicit ctx: Context): Option[(Type, Type, Type)] =
+        xs.collect {
+          case func1 @ RefinedType(RefinedType(parent, _, t1), _, r)
+            if func1.isRef(Function1) => (t1, r, func1)
+        }.headOption
+
+      collectFunc1(cref.info.parentsWithArgs).flatMap { case (t1, r, func1) =>
+        if (
+          !argTypes.contains(t1.typeSymbol) ||
+          !returnTypes.contains(r.typeSymbol) ||
+          blacklisted.exists(sym => cref.symbol.derivesFrom(sym))
+        ) None
+        else Some((cref, t1, r, func1))
+      }
+    }
+  }
+
+  private object ShouldTransformTree {
+    def unapply(tree: Template)(implicit ctx: Context): Option[(Tree, Type, Type)] =
+      tree.parents
+      .map { t => (t.tpe, t) }
+      .collect {
+        case (tp @ RefinedType(RefinedType(parent, _, t1), _, r), func1)
+        if tp.isRef(Function1) &&
+           argTypes.contains(t1.typeSymbol) &&
+           returnTypes.contains(r.typeSymbol) => (func1, t1, r)
+      }
+      .headOption
+  }
+
+  private def specializedName(name: Name, t1: Type, r: Type)(implicit ctx: Context): Name =
+    name.specializedFor(List(t1, r), List(t1, r).map(_.typeSymbol.name))
+
+  def transformDenot(cref: ClassDenotation, t1: Type, r: Type, func1: Type)(implicit ctx: Context): SingleDenotation = {
+    val specializedFunction: TypeRef =
+      ctx.requiredClassRef(functionPkg ++ specializedName(functionName, t1, r))
```

Comment:
`requiredClassRef` would fail compilation if there's no such class. I don't think it should be used here.